core/src/replay_stage.rs

```diff
@@ -3509,6 +3509,13 @@ impl ReplayStage {
         };
 
         if let Some((bank, switch_fork_decision)) = selected_fork {
+            let check_slot = if switch_fork_decision.can_vote() {
```

Comment:
nit: check_slot sounds ambiguous, maybe slot_for_logging or something along that line.

Reply:
the slot we check is the one we log. if we can vote then we check the slot we want to vote on. If we can't vote we should check the slot we *wanted* to vote on, because there's no point in checking the reset slot (produces useless and confusing info)

For example if you're looking to switch from slot 100 to a different fork slot 102 you could see something like this:
`HeaviestForkFailures::FailedSwitchThreshold(...), HeaviestForkFailures::FailedThreshold(...), HeaviestForkFailures::NoPropagatedConfirmed`
In this case the switch threshold failure is for slot 102, but the threshold and propagation check are actually being performed on the reset bank descending from 100 which is useless information and confusing for someone debugging. 

Gets even more confusing when your last vote is for a duplicate fork, as your reset bank could be not descending for your last vote. So you get metrics telling you:
`HeaviestForkFailures::FailedSwitchThreshold, HeaviestForkFailures::LockedOut`
Which implies that you can't switch yet because you need proof stake + you're locked out, but in fact you don't know if you're locked out because you are checking lockout for the reset bank instead.

Reply:
Hmm this logic is getting too convoluted because the outputs of this match statement are a bit messy.

I think we can clean this logic up so that we avoid this logging problem altogether, i.e. if we get the switch failure, we don't even need to do the remaining checks.

I propose a cleanup of this function here: https://github.com/solana-labs/solana/pull/33421